source/common/stream_info/utility.cc

```diff
@@ -43,6 +49,231 @@ absl::optional<ResponseFlag> ResponseFlagUtils::toResponseFlag(absl::string_view
+  case ProxyStatusError::HttpResponseTrailerSectionSize:
+    return HTTP_RESPONSE_TRAILER_SECTION_SIZE;
+  case ProxyStatusError::HttpResponseTrailerSize:
+    return HTTP_RESPONSE_TRAILER_SIZE;
+  case ProxyStatusError::HttpResponseTransferCoding:
+    return HTTP_RESPONSE_TRANSFER_CODING;
+  case ProxyStatusError::HttpResponseContentCoding:
+    return HTTP_RESPONSE_CONTENT_CODING;
+  case ProxyStatusError::HttpResponseTimeout:
+    return HTTP_RESPONSE_TIMEOUT;
+  case ProxyStatusError::HttpUpgradeFailed:
+    return HTTP_UPGRADE_FAILED;
+  case ProxyStatusError::HttpProtocolError:
+    return HTTP_PROTOCOL_ERROR;
+  case ProxyStatusError::ProxyInternalResponse:
+    return PROXY_INTERNAL_RESPONSE;
+  case ProxyStatusError::ProxyInternalError:
+    return PROXY_INTERNAL_ERROR;
+  case ProxyStatusError::ProxyConfigurationError:
+    return PROXY_CONFIGURATION_ERROR;
+  case ProxyStatusError::ProxyLoopDetected:
+    return PROXY_LOOP_DETECTED;
+  }
+  return "-";
+}
+
+const absl::optional<ProxyStatusError>
+ProxyStatusUtils::fromStreamInfo(const StreamInfo& stream_info) {
+  // NB: This mapping from Envoy-specific ResponseFlag enum to Proxy-Status
+  // error enum is lossy, since ResponseFlag is really a bitset of many
+  // ResponseFlag enums. Here, we search the list of all known ResponseFlag values in
+  // enum order, returning the first matching ProxyStatusError.
+  if (stream_info.hasResponseFlag(ResponseFlag::FailedLocalHealthCheck)) {
+    return ProxyStatusError::DestinationUnavailable;
+  } else if (stream_info.hasResponseFlag(ResponseFlag::NoHealthyUpstream)) {
+    return ProxyStatusError::DestinationUnavailable;
+  } else if (stream_info.hasResponseFlag(ResponseFlag::UpstreamRequestTimeout)) {
+    return ProxyStatusError::ConnectionTimeout;
+  } else if (stream_info.hasResponseFlag(ResponseFlag::LocalReset)) {
+    return ProxyStatusError::ConnectionTimeout;
+  } else if (stream_info.hasResponseFlag(ResponseFlag::UpstreamRemoteReset)) {
+    return ProxyStatusError::ConnectionTerminated;
+  } else if (stream_info.hasResponseFlag(ResponseFlag::UpstreamConnectionFailure)) {
+    return ProxyStatusError::ConnectionRefused;
+  } else if (stream_info.hasResponseFlag(ResponseFlag::UpstreamConnectionTermination)) {
+    return ProxyStatusError::ConnectionTerminated;
+  } else if (stream_info.hasResponseFlag(ResponseFlag::UpstreamOverflow)) {
+    return ProxyStatusError::ConnectionLimitReached;
+  } else if (stream_info.hasResponseFlag(ResponseFlag::NoRouteFound)) {
+    return ProxyStatusError::DestinationNotFound;
+  } else if (stream_info.hasResponseFlag(ResponseFlag::RateLimited)) {
+    return ProxyStatusError::ConnectionLimitReached;
+  } else if (stream_info.hasResponseFlag(ResponseFlag::RateLimitServiceError)) {
+    return ProxyStatusError::ConnectionLimitReached;
+  } else if (stream_info.hasResponseFlag(ResponseFlag::UpstreamRetryLimitExceeded)) {
+    return ProxyStatusError::ConnectionTerminated;
```

Comment:
I think DestinationUnavailable is more appropriate here.

Reply:
Done.